semver.md

```diff
@@ -286,6 +286,19 @@ modification depends on whether you updated your dependencies in order to fix
 a bug or introduce new functionality. I would usually expect additional code
 for the latter instance, in which case it's obviously a minor level increment.
 
+#### Special note for JVM or similar environments
```

Comment:
That makes it sound like it's a special case when it's actually the norm. It's problematic even in environments that *do* allow multiple versions of the same library.

I think the real issue is that the previous FAQ is misleading.

For example, let's take a really simple API:

```frobnicate: Widget -> Result```

Now upgrade the widget library to an incompatible version. (Eg: 1.0.0 to 2.0.0) Now the API is:

```frobnicate: Widget -> Result```

No change, right? Except this API is incompatible with the previous API because it now requires a v2 Widget instead of a v1 Widget.

Of course anyone reading the previous FAQ *carefully enough* would realize that when the version of a library is *implicitly* part of the public API, then changing it *is* changing the public API despite `diff` not showing a difference. Many people appear to not read it carefully enough.

And also of course, in environments like the JVM that only allow a single version of a library, even when a library is used "purely internally" and not even mentioned in the public API, it is still part of the *undocumented* public API in that it precludes the use of a different version of that library.

Reply:
@aij, you are getting interface and package mixed up. An interface never has any implementation issues to consider.

Reply:
@jwdonahue Could you elaborate? I didn't think I was talking about packages at all.

I was talking about non-textual (contextual?) interface changes. An interface can be changed in an incompatible way even without any textual changes to it's definition, by merely changing the context. For the above example, if the `Widget` interface is changed incompatibly, it changes the meaning of `Widget -> Result`, even though that text itself is not changed.

The reason I put quotes around "purely internally", is because a lot of environments don't support making use/nonuse of a library into an implementation detail. It necessarily becomes part of the public interface in environments that can't/don't link in multiple copies of the same library. (JVM, *nix shared libraries, most (all?) C static libraries, OCaml, Python, etc)

Reply:
@aij, I must have misread something. I thought you were talking about interfaces when in fact you were talking about packages/libraries.